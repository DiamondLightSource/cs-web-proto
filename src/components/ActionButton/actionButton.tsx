import React from "react";
import { Actions, executeActions } from "../../actions";
import { InferWidgetProps, PVWidget, PVWidgetPropType } from "../Widget/widget";
import classes from "./actionButton.module.css";

export interface ActionButtonProps {
  text: string;
  onClick: (event: React.MouseEvent<HTMLButtonElement>) => void;
  style?: {};
  image?: string;
}

export const ActionButtonComponent = (
  props: ActionButtonProps
): JSX.Element => {
  if (props.image !== undefined) {
    return (
      <button className={classes.image} onClick={props.onClick}>
        <img src={props.image} alt={props.image}></img>
        <br></br>
        {props.text}
      </button>
    );
  } else {
    return <button onClick={props.onClick}>{props.text}</button>;
  }
};

export interface ActionButtonWidgetProps {
  text: string;
  actions: Actions;
  style?: {};
  image?: string;
}

// Menu button which also knows how to write to a PV
export const ActionButtonWidget = (
  props: ActionButtonWidgetProps
): JSX.Element => {
  // Function to send the value on to the PV
  function onClick(event: React.MouseEvent<HTMLButtonElement>): void {
    if (props.actions !== undefined) executeActions(props.actions);
  }
  return (
    <ActionButtonComponent
      text={props.text}
      style={props.style}
      onClick={onClick}
      image={props.image}
    />
  );
};

export const ActionButton = (
  props: InferWidgetProps<typeof PVWidgetPropType>
): JSX.Element => <PVWidget baseWidget={ActionButtonWidget} {...props} />;

ActionButton.propTypes = PVWidgetPropType;
