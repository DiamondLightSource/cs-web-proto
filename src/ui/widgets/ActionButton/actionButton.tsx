import React, { useContext } from "react";
import { WidgetActions, executeActions } from "../widgetActions";
import { Widget } from "../widget";
import { PVWidgetPropType } from "../widgetProps";
import classes from "./actionButton.module.css";
import { useHistory } from "react-router-dom";
import { registerWidget } from "../register";
import {
  ActionsPropType,
  StringProp,
  StringPropOpt,
  InferWidgetProps,
  ColorPropOpt,
  FontPropOpt,
  BorderPropOpt,
} from "../propTypes";
import { BaseUrlContext } from "../../../baseUrl";
import { Color } from "../../../types/color";
import { Font } from "../../../types/font";
import { Border } from "../../../types/border";

export interface ActionButtonProps {
  text: string;
  onClick: (event: React.MouseEvent<HTMLButtonElement>) => void;
  image?: string;
  backgroundColor?: Color;
  foregroundColor?: Color;
  border?: Border;
  font?: Font;
}

export const ActionButtonComponent = (
  props: ActionButtonProps
): JSX.Element => {
  const baseUrl = useContext(BaseUrlContext);
  let src = props.image;
  if (src !== undefined && !src?.startsWith("http")) {
    src = `${baseUrl}/img/${src}`;
  }
  return (
    <button
      className={classes.actionbutton}
      onClick={props.onClick}
      style={{
        backgroundColor: props.backgroundColor?.rgbaString(),
        color: props.foregroundColor?.rgbaString(),
        ...props.font?.css(),
        ...props.border?.css(),
      }}
    >
      {src !== undefined ? (
        <figure className={classes.figure}>
          <img src={src} alt={props.image}></img>
          <figcaption>{props.text}</figcaption>
        </figure>
      ) : (
        <span>{props.text}</span>
      )}
    </button>
  );
};

const ActionButtonPropType = {
  text: StringProp,
  actions: ActionsPropType,
  image: StringPropOpt,
  backgroundColor: ColorPropOpt,
  foregroundColor: ColorPropOpt,
  font: FontPropOpt,
  border: BorderPropOpt,
};

const ActionButtonProps = {
  ...ActionButtonPropType,
  ...PVWidgetPropType,
};

// Menu button which also knows how to write to a PV
export const ActionButtonWidget = (
  props: InferWidgetProps<typeof ActionButtonPropType>
): JSX.Element => {
  // Function to send the value on to the PV
  const history = useHistory();
  function onClick(event: React.MouseEvent<HTMLButtonElement>): void {
    if (props.actions !== undefined)
      executeActions(props.actions as WidgetActions, history);
  }
  return (
    <ActionButtonComponent
      text={props.text}
      onClick={onClick}
      image={props.image}
      backgroundColor={props.backgroundColor}
      foregroundColor={props.foregroundColor}
      font={props.font}
      border={props.border}
    />
  );
};

export const ActionButton = (
  props: InferWidgetProps<typeof ActionButtonProps>
): JSX.Element => <Widget baseWidget={ActionButtonWidget} {...props} />;

registerWidget(ActionButton, ActionButtonProps, "actionbutton");
