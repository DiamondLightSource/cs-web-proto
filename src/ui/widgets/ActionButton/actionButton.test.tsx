import React from "react";
import { ActionButtonComponent } from "./actionButton";
import { create } from "react-test-renderer";
import { fireEvent, render, waitFor } from "@testing-library/react";

const mock = jest.fn();
const actionButton = <ActionButtonComponent text={"hello"} onClick={mock} />;

describe("<ActionButton />", (): void => {
  test("it matches the snapshot", (): void => {
    const snapshot = create(actionButton);
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test("it renders a button", (): void => {
    const { getByRole } = render(actionButton);
    const button = getByRole("button") as HTMLButtonElement;
    expect(button.textContent).toEqual("hello");
  });

  test("function called on click", async (): Promise<void> => {
    const { getByRole } = render(actionButton);
    const button = getByRole("button") as HTMLButtonElement;
    fireEvent.click(button);
    await waitFor(() => expect(mock).toHaveBeenCalled());
  });
});
