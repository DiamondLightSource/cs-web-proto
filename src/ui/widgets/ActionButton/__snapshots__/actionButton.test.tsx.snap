// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionButton /> it matches the snapshot 1`] = `
<button
  className="actionbutton"
  onClick={[Function]}
  style={
    Object {
      "backgroundColor": undefined,
      "color": undefined,
      "outline": undefined,
      "outlineOffset": undefined,
      "overflow": "hidden",
      "visibility": undefined,
    }
  }
>
  <span>
    hello
  </span>
</button>
`;
